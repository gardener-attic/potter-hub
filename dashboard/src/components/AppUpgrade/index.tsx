import AppUpgrade from "./AppUpgrade";

export default AppUpgrade;
