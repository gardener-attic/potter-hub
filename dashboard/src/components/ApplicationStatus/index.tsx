import ApplicationStatus from "./ApplicationStatus";

export default ApplicationStatus;
