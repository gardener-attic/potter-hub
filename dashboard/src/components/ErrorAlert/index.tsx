import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
