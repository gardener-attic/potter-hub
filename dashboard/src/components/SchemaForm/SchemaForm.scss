.SchemaForm > .form-group > fieldset {
  padding: 0;
  border: none;
}
