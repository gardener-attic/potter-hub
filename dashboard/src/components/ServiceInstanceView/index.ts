import ServiceInstanceView from "./ServiceInstanceView";

export default ServiceInstanceView;
