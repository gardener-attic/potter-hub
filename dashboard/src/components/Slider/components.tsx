// CODE FROM: https://sghall.github.io/react-compound-slider/#/slider-demos/horizontal
import * as React from "react";
import { GetHandleProps, GetTrackProps, SliderItem } from "react-compound-slider";

// *******************************************************
// HANDLE COMPONENT
// *******************************************************
interface IHandleProps {
  domain: number[];
  handle: SliderItem;
  getHandleProps: GetHandleProps;
}

export const Handle: React.SFC<IHandleProps> = ({
  domain: [min, max],
  handle: { id, value, percent },
  getHandleProps,
}) => (
  <div
    role="slider"
    aria-valuemin={min}
    aria-valuemax={max}
    aria-valuenow={value}
    style={{
      left: `${percent}%`,
      position: "absolute",
      marginLeft: "-11px",
      marginTop: "-6px",
      zIndex: 2,
      width: 24,
      height: 24,
      cursor: "pointer",
      borderRadius: "50%",
      boxShadow: "1px 1px 1px 1px rgba(0, 0, 0, 0.2)",
      backgroundColor: "#34568f",
    }}
    {...getHandleProps(id)}
  />
);

// *******************************************************
// TRACK COMPONENT
// *******************************************************
interface ITrackProps {
  source: SliderItem;
  target: SliderItem;
  getTrackProps: GetTrackProps;
}

export const Track: React.SFC<ITrackProps> = ({ source, target, getTrackProps }) => (
  <div
    style={{
      position: "absolute",
      height: 14,
      zIndex: 1,
      backgroundColor: "#7aa0c4",
      borderRadius: 7,
      cursor: "pointer",
      left: `${source.percent}%`,
      width: `${target.percent - source.percent}%`,
    }}
    {...getTrackProps()}
  />
);
