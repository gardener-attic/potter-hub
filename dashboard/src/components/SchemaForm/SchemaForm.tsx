import { JSONSchema6 } from "json-schema";
import * as React from "react";
import Form, { ISubmitEvent } from "react-jsonschema-form";

import ArrayFieldTemplate from "./ArrayFieldTemplate";
import CustomObjectField from "./CustomObjectField";
import FieldTemplate from "./FieldTemplate";
import "./SchemaForm.css";

interface ISchemaFormProps {
  schema: JSONSchema6;
  onSubmit?: (result: ISubmitEvent<any>) => void;
}

class SchemaForm extends React.Component<ISchemaFormProps> {
  public render() {
    return (
      <Form
        className="SchemaForm"
        schema={this.props.schema}
        onSubmit={this.props.onSubmit}
        autocomplete="off"
        FieldTemplate={FieldTemplate}
        ArrayFieldTemplate={ArrayFieldTemplate}
        fields={{ ObjectField: CustomObjectField }}
        noValidate={true}
      >
        {this.props.children}
      </Form>
    );
  }
}

export default SchemaForm;
