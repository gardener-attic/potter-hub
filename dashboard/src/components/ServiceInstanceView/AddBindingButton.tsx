import * as React from "react";
import Modal from "react-modal";

import { JSONSchema6 } from "json-schema";
import { ISubmitEvent } from "react-jsonschema-form";
import { IRBACRole } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import SchemaForm from "../SchemaForm";

interface IAddBindingButtonProps {
  disabled: boolean;
  error?: Error;
  instanceRefName: string;
  namespace: string;
  addBinding: (
    bindingName: string,
    instanceName: string,
    namespace: string,
    parameters: {},
  ) => Promise<boolean>;
  onAddBinding: () => void;
  bindingSchema?: JSONSchema6;
}

interface IAddBindingButtonState {
  modalIsOpen: boolean;
  // deployment options
  bindingName: string;
  displayNameForm: boolean;
  // Name of the binding that was submitted for creation
  // This is different than bindingName since it is also used in the error banner
  // and we do not want to use bindingName since it is controller by the form field.
  latestSubmittedReleaseName: string;
}

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "servicecatalog.k8s.io",
    resource: "servicebindings",
    verbs: ["create"],
  },
];

class AddBindingButton extends React.Component<IAddBindingButtonProps, IAddBindingButtonState> {
  public state = {
    bindingName: `${this.props.instanceRefName}-binding`,
    displayNameForm: true,
    modalIsOpen: false,
    latestSubmittedReleaseName: `${this.props.instanceRefName}-binding`,
  };

  public render() {
    let { bindingSchema } = this.props;
    const { modalIsOpen, displayNameForm } = this.state;
    if (!bindingSchema || !(bindingSchema.$ref || bindingSchema.properties)) {
      bindingSchema = {
        properties: {
          kubeappsRawParameters: {
            title: "Parameters",
            type: "object",
          },
        },
        type: "object",
      };
    }
    return (
      <div className="AddBindingButton">
        <button
          className="button button-primary"
          onClick={this.openModal}
          disabled={this.props.disabled}
        >
          Add Binding
        </button>
        <Modal isOpen={modalIsOpen} onRequestClose={this.closeModal}>
          {this.props.error && (
            <ErrorSelector
              error={this.props.error}
              resource={`Service Binding "${this.state.latestSubmittedReleaseName}"`}
              action="create"
              namespace={this.props.namespace}
              defaultRequiredRBACRoles={{ create: RequiredRBACRoles }}
            />
          )}
          <div className="bind-form">
            <h1>Add Binding</h1>
            {displayNameForm ? (
              <SchemaForm schema={this.nameSchema()} onSubmit={this.handleNameChange}>
                <div>
                  <button className="button button-primary" type="submit">
                    Continue
                  </button>
                  <button className="button" onClick={this.closeModal}>
                    Cancel
                  </button>
                </div>
              </SchemaForm>
            ) : (
              <SchemaForm schema={bindingSchema} onSubmit={this.handleBind}>
                <div>
                  <button className="button button-primary" type="submit">
                    Submit
                  </button>
                  <button className="button" onClick={this.handleBackButton}>
                    Back
                  </button>
                </div>
              </SchemaForm>
            )}
          </div>
        </Modal>
      </div>
    );
  }

  private closeModal = () => this.setState({ modalIsOpen: false });
  private openModal = () => this.setState({ modalIsOpen: true });
  private handleNameChange = ({ formData }: ISubmitEvent<{ Name: string }>) => {
    this.setState({ bindingName: formData.Name, displayNameForm: false });
  };
  private handleBackButton = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    this.setState({ displayNameForm: true });
  };

  private handleBind = async ({ formData }: ISubmitEvent<{ kubeappsRawParameters: {} }>) => {
    const { kubeappsRawParameters, ...rest } = formData;
    this.setState({ latestSubmittedReleaseName: this.state.bindingName });
    const added = await this.props.addBinding(
      this.state.bindingName,
      this.props.instanceRefName,
      this.props.namespace,
      kubeappsRawParameters || rest,
    );
    if (added) {
      this.closeModal();
      this.props.onAddBinding();
    }
  };

  private nameSchema(): JSONSchema6 {
    return {
      properties: {
        Name: {
          default: this.state.bindingName,
          description: "Name for ServiceBinding",
          type: "string",
        },
      },
      required: ["Name"],
      type: "object",
    };
  }
}

export default AddBindingButton;
