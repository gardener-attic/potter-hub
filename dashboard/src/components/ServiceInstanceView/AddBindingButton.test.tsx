import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import { JSONSchema6 } from "json-schema";
import { ErrorSelector } from "../../components/ErrorAlert";
import SchemaForm from "../../components/SchemaForm";
import AddBindingButton from "./AddBindingButton";

const defaultName = "my-class";
const defaultNS = "default";
const defaultProps = {
  disabled: false,
  instanceRefName: defaultName,
  namespace: defaultNS,
  addBinding: jest.fn(),
  onAddBinding: jest.fn(),
};

it("shows a button", () => {
  const wrapper = shallow(<AddBindingButton {...defaultProps} />);
  wrapper.setState({ isProvisioning: true });
  expect(wrapper.find(".button")).toExist();
});

it("disables the button", () => {
  const wrapper = shallow(<AddBindingButton {...defaultProps} disabled={true} />);
  wrapper.setState({ isProvisioning: true });
  expect(
    wrapper
      .find(".button")
      .filterWhere(e => e.text() === "Add Binding")
      .prop("disabled"),
  ).toBe(true);
});

context("when the modal is open", () => {
  it("shows an error if is present", () => {
    const wrapper = shallow(<AddBindingButton {...defaultProps} error={new Error()} />);
    wrapper.setState({ modalIsOpen: true });

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("should display the name form", () => {
    const wrapper = shallow(<AddBindingButton {...defaultProps} />);
    wrapper.setState({ modalIsOpen: true, displayNameForm: true });

    const form = wrapper.find(SchemaForm);
    expect(form).toExist();
    expect((form.prop("schema") as any).properties).toHaveProperty("Name");
  });

  it("should display the default schema if it's not in the plan spec", () => {
    const wrapper = shallow(<AddBindingButton {...defaultProps} bindingSchema={undefined} />);
    wrapper.setState({ modalIsOpen: true, displayNameForm: false });

    const form = wrapper.find(SchemaForm);
    expect(form).toExist();
    const schema = form.prop("schema") as any;
    expect(schema).toMatchObject({
      properties: {
        kubeappsRawParameters: {
          title: "Parameters",
          type: "object",
        },
      },
      type: "object",
    });
  });

  it("should display the spec schema", () => {
    const newSchema = {
      properties: {
        kubeappsRawParameters: {
          title: "Foo",
          type: "string",
        },
      },
      type: "object",
    } as JSONSchema6;
    const wrapper = shallow(<AddBindingButton {...defaultProps} bindingSchema={newSchema} />);
    wrapper.setState({ modalIsOpen: true, displayNameForm: false });

    const form = wrapper.find(SchemaForm);
    expect(form).toExist();
    const schema = form.prop("schema") as any;
    expect(schema).toEqual(newSchema);
  });
});
