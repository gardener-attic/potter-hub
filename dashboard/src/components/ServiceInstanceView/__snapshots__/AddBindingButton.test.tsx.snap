// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the modal is open shows an error if is present 1`] = `
<div
  className="AddBindingButton"
>
  <button
    className="button button-primary"
    disabled={false}
    onClick={[Function]}
  >
    Add Binding
  </button>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    isOpen={true}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <ErrorSelector
      action="create"
      defaultRequiredRBACRoles={
        Object {
          "create": Array [
            Object {
              "apiGroup": "servicecatalog.k8s.io",
              "resource": "servicebindings",
              "verbs": Array [
                "create",
              ],
            },
          ],
        }
      }
      error={[Error]}
      namespace="default"
      resource="Service Binding \\"my-class-binding\\""
    />
    <div
      className="bind-form"
    >
      <h1>
        Add Binding
      </h1>
      <SchemaForm
        onSubmit={[Function]}
        schema={
          Object {
            "properties": Object {
              "Name": Object {
                "default": "my-class-binding",
                "description": "Name for ServiceBinding",
                "type": "string",
              },
            },
            "required": Array [
              "Name",
            ],
            "type": "object",
          }
        }
      >
        <div>
          <button
            className="button button-primary"
            type="submit"
          >
            Continue
          </button>
          <button
            className="button"
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>
      </SchemaForm>
    </div>
  </Modal>
</div>
`;
